### PR TITLE
Implement hard limit

### PR DESCRIPTION
**Pull request by maciekpac**
### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [x] Code compiles without errors
- [x] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [x] Created tests which will fail without the change (if possible)
- [ ] Extended the README/documentation (if necessary)
- [] All newly added files have proprietary license (if necessary)
- [ ] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/789)

### CODE REVIEW

**Review by PatKamin — CHANGES_REQUESTED** [review]
> Reviewed 5 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: 5 of 16 files reviewed, 2 unresolved discussions (waiting on @bratpiorka and @maciekpac)

___
*[include/memkind/internal/bigary.h, line 42 at r1](https://reviewable.io/reviews/memkind/memkind/789#-MzYk1-l1-3gUt0uwBRT:-MzYk1-l1-3gUt0uwBRU:btawsqj) ([raw file](https://github.com/memkind/memkind/blob/9783f3de342f513b40406b1eaf752d1556839c79/include/memkind/internal/bigary.h#L42)):*
> ```objc
>                              int __flags, int __fd, __off_t __offset)
> {
>     (void)arg;
> ```

What is the purpose of providing this argument?
___
*[include/memkind/internal/mtt_internals.h, line 13 at r1](https://reviewable.io/reviews/memkind/memkind/789#-MzYl-XqDyy9gW6u4CDc:-MzYl-XqDyy9gW6u4CDd:b-ng0qk9) ([raw file](https://github.com/memkind/memkind/blob/9783f3de342f513b40406b1eaf752d1556839c79/include/memkind/internal/mtt_internals.h#L13)):*
> ```objc
> #ifdef __cplusplus
> #include <atomic>
> // HACK
> ```

If you see other way this can be handled, please add TODO. Otherwise this comment isn't useful.

File: src/mtt_internals.c [hunk 1]
```diff
@@ -125,9 +163,71 @@ MEMKIND_EXPORT void mtt_internals_touch(MttInternals *internals,
     ranking_touch(internals->pmemRanking, address);
 }
 
+MEMKIND_EXPORT size_t mtt_internals_ranking_balance(MttInternals *internals,
+                                                    atomic_size_t *dram_size,
```
Comment (PatKamin) [inline src/mtt_internals.c:167]: Perhaps this would better describe this var: `dram_size` -> `used_dram`

**Review by maciekpac — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: 5 of 16 files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[include/memkind/internal/bigary.h, line 42 at r1](https://reviewable.io/reviews/memkind/memkind/789#-MzYk1-l1-3gUt0uwBRT:-MzZJI2g-pax8NwXB8t1:b-lq3y65) ([raw file](https://github.com/memkind/memkind/blob/9783f3de342f513b40406b1eaf752d1556839c79/include/memkind/internal/bigary.h#L42)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

What is the purpose of providing this argument?
</blockquote></details>

This function is used as a callback and has to match the callback signature.

In other places, we need to check how much dram is currently used and increase that value right before doing an mmap. This argument is there to do exactly that, without creating circular dependency.

Without callback, by directly calling functions:

mtt_allocator->internals->pool->slab->bigary->mtt_allocator

with callback and with argument:

mtt_allocator->internals->pool->slab->bigary

This circular dependency might seriously complicate things, such as overloading mmap.
___
*[include/memkind/internal/mtt_internals.h, line 13 at r1](https://reviewable.io/reviews/memkind/memkind/789#-MzYl-XqDyy9gW6u4CDc:-MzZKFoZ85pjv6u4IGFM:bb6un1j) ([raw file](https://github.com/memkind/memkind/blob/9783f3de342f513b40406b1eaf752d1556839c79/include/memkind/internal/mtt_internals.h#L13)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

If you see other way this can be handled, please add TODO. Otherwise this comment isn't useful.
</blockquote></details>

Done
___
*[src/mtt_internals.c, line 167 at r1](https://reviewable.io/reviews/memkind/memkind/789#-MzV8NMkvMdCCYCdsdwo-r1-167:-MzZKWf32NLb-4m6G7S6:bb6un1j) ([raw file](https://github.com/memkind/memkind/blob/9783f3de342f513b40406b1eaf752d1556839c79/src/mtt_internals.c#L167)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Perhaps this would better describe this var: `dram_size` -> `used_dram`
</blockquote></details>

Done

**Review by PatKamin — COMMENTED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: all files reviewed, 1 unresolved discussion (waiting on @bratpiorka and @PatKamin)

**Review by PatKamin — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: all files reviewed, 2 unresolved discussions (waiting on @bratpiorka, @maciekpac, and @PatKamin)

___
*[src/mtt_allocator.c, line 13 at r3](https://reviewable.io/reviews/memkind/memkind/789#-MznWIiSErz8dNeoPUvQ:-MznWIiT0_zqp_6reYF2:b-ynzrkk) ([raw file](https://github.com/memkind/memkind/blob/f53e6eaa5d9cb64c9d9eeee146903224b4ac71af/src/mtt_allocator.c#L13)):*
> ```C
> #include <stdatomic.h>
> #include <stdlib.h>
> #include <threads.h>
> ```

Add #ifdef HAVE_THREADS_H like in memtier.c

**Review by maciekpac — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: 15 of 16 files reviewed, 2 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[src/mtt_allocator.c, line 13 at r3](https://reviewable.io/reviews/memkind/memkind/789#-MznWIiSErz8dNeoPUvQ:-Mzo5uM5EfCqeBWo28N4:b-459bvr) ([raw file](https://github.com/memkind/memkind/blob/f53e6eaa5d9cb64c9d9eeee146903224b4ac71af/src/mtt_allocator.c#L13)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Add #ifdef HAVE_THREADS_H like in memtier.c
</blockquote></details>

This header is not needed in the final version; removed

**Review by bratpiorka — CHANGES_REQUESTED** [review]
> Reviewed 5 of 16 files at r1, 7 of 10 files at r2, 2 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: all files reviewed, 6 unresolved discussions (waiting on @maciekpac and @PatKamin)

___
*[src/mtt_allocator.c, line 45 at r5](https://reviewable.io/reviews/memkind/memkind/789#-MzomGDVDmSjWY3LB9G0:-MzomGDVDmSjWY3LB9G1:b-mo7fq5) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/src/mtt_allocator.c#L45)):*
> ```C
> // static function definitions ------------------------------------------------
> 
> /// @brief Wrapper for mmap
> ```

maybe "wrapper for mmap with DRAM/PMEM balancing - at the function exit we should have allocated at most hardLimit data in DRAM" or something like that?
___
*[src/mtt_internals.c, line 224 at r5](https://reviewable.io/reviews/memkind/memkind/789#-Mzomknc28Z_qPq4x4Ot:-Mzomknc28Z_qPq4x4Ou:b-b2tms0) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/src/mtt_internals.c#L224)):*
> ```C
>         assert((int64_t)*used_dram >= 0 && "internal error, moved too much");
>     } else if (temp_dram_size + TRACED_PAGESIZE <
>                internals->limits
> ```

why there is a line break?
___
*[test/memkind_memtier_data_movement_test.cpp, line 757 at r5](https://reviewable.io/reviews/memkind/memkind/789#-MzolS7t7rcVTEq1qhqQ:-MzolS7t7rcVTEq1qhqR:b-2ox9x4) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/test/memkind_memtier_data_movement_test.cpp#L757)):*
> ```cpp
>     mtt_allocator_create(&mtt_allocator, &limits);
>     // metadata overhead for mtt_allocator
>     const size_t metadata_size = 3 * BIGARY_PAGESIZE;
> ```

please add a comment why "3 * BIGARY_PAGESIZE" 
___
*[test/memkind_memtier_data_movement_test.cpp, line 853 at r5](https://reviewable.io/reviews/memkind/memkind/789#-Mzom4gH54I_FogbgvDy:-Mzom4gH54I_FogbgvDz:bjuznpe) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/test/memkind_memtier_data_movement_test.cpp#L853)):*
> ```cpp
>         tab[i] = (int *)mtt_allocator_malloc(&mtt_allocator, size);
> 
>     size_t dram_ranking_size =
> ```

don't we need here any kind of waiting for bg thread to finish moving pages to PMEM?

**Review by maciekpac — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: 13 of 16 files reviewed, 6 unresolved discussions (waiting on @bratpiorka and @PatKamin)

___
*[src/mtt_allocator.c, line 45 at r5](https://reviewable.io/reviews/memkind/memkind/789#-MzomGDVDmSjWY3LB9G0:-MzorLQn11gEHYzalEWw:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/src/mtt_allocator.c#L45)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

maybe "wrapper for mmap with DRAM/PMEM balancing - at the function exit we should have allocated at most hardLimit data in DRAM" or something like that?
</blockquote></details>

Done.
___
*[src/mtt_internals.c, line 224 at r5](https://reviewable.io/reviews/memkind/memkind/789#-Mzomknc28Z_qPq4x4Ot:-MzoySiM3Qw2_4xMOr_N:bx90onf) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/src/mtt_internals.c#L224)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

why there is a line break?
</blockquote></details>

Because of the TODO comment; I removed the TODO  because it was no longer valid and everything looks OK now.
___
*[test/memkind_memtier_data_movement_test.cpp, line 757 at r5](https://reviewable.io/reviews/memkind/memkind/789#-MzolS7t7rcVTEq1qhqQ:-Mzp22D08jcS7AtOmCAT:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/test/memkind_memtier_data_movement_test.cpp#L757)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

please add a comment why "3 * BIGARY_PAGESIZE" 
</blockquote></details>

Done.
___
*[test/memkind_memtier_data_movement_test.cpp, line 853 at r5](https://reviewable.io/reviews/memkind/memkind/789#-Mzom4gH54I_FogbgvDy:-Mzp2dYz1wS_cGovq61W:b-zexsh0) ([raw file](https://github.com/memkind/memkind/blob/87cfdeeb8bf5660fe77d66531ced07083865ea21/test/memkind_memtier_data_movement_test.cpp#L853)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

don't we need here any kind of waiting for bg thread to finish moving pages to PMEM?
</blockquote></details>

we are making checks for hard limit, which should be valid even before pages are moved to PMEM - this is the whole point... unfortunately, we currently don't have a function to disable async PMEM->DRAM movement to make a better check...

I added some more checks

**Review by bratpiorka — APPROVED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: all files reviewed, 2 unresolved discussions (waiting on @PatKamin)

**Review by bratpiorka — APPROVED** [review]
> Reviewed 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @maciekpac)

**Review by PatKamin — APPROVED** [review]
> Reviewed 1 of 1 files at r5, 2 of 3 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/789)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @maciekpac)

**Pull request merged**